Introduce Zephyr Filesystem VFS interface

VynDragon opened the pull request:
### Summary

Adds VFS interface for Zephyr filesystems

This allows using Zephyr's supported VFS and interface to them via MPY, making mpy side VFS unecessary (eg you can disable LFS2 in mpy but still do LFS2 because it's using zephyr's)


### Testing

Basic tests on nrf52840, would recommend thorough testing.

15/08/2025: also tested on rp2040.

VynDragon committed:
zephyr: Introduce Zephyr native filesystem VFS interface.

This allows using Zephyr's supported VFS and interface it to MicroPython's
VFS.  That makes the MicroPython-side filesystem drivers unecessary, eg you
can disable LFS2 in MicroPython but still do LFS2 because it's using
Zephyr's.

An advantage of this is that the filesystem can be shared between Zephyr C
code and MicroPython.

Signed-off-by: Vdragon <mail@massdriver.space>

jonnor commented:
What is needed to set this up on a device? Any limitations, or can one use any file-system operations from Python like on other ports?

VynDragon commented:
Basically you need to setup a fstap and then format it: https://docs.zephyrproject.org/latest/services/file_system/index.html

After that it works like any mpy fs.

VynDragon commented:
rebased on main

VynDragon commented:
> Since there's already code to mount a filesystem on the partition labeled `storage` as `/flash` (I did add some code to my PR #17679 to create a filesystem if the mount fails), I guess this is mostly useful for other partitions?

No, this is a completely different method as the existing one, and your PR doesnt even use zephyr lfs2, you could disable it in your code and it would still work.

dpgeorge commented:
This is a nice addition, but I have a few general comments:
1. It's going to get confusing for new users which filesystem to use (native zephyr vs MicroPython).  I don't think there's any way around that except through documentation and examples.
2. You also need to explicitly enable this, and maybe worth adding something to the zephyr README to show how to do that.
3. This won't be tested by CI so most likely will accrue bugs.  Zephyr CI is already slow, so maybe one of the existing CI builds (eg `mimxrt1050_evk`) could have a custom `CONF_FILE` specified for the CI to test building this FS code?

dpgeorge commented on ports/zephyr/zephyr_fs.c:
```diff
@@ -0,0 +1,767 @@
+    return path;
+}
+
+typedef struct _pyb_file_obj_t {
```
please rename thing struct to something like `zephyr_file_obj_t`

VynDragon committed:
zephyr/CMakeLists.txt: Fix FS config options being ignored.

This makes CONFIG_MICROPY_VFS_* actually do things.

Signed-off-by: Vdragon <mail@massdriver.space>

VynDragon committed:
zephyr/boards: Setup rpi_pico to use Zephyr FS instead of MPY FS.

Use new Zephyr FS, and be a demonstration of that configuration.

This is still on-disk compatible with existing filesystems on the internal
flash.

Signed-off-by: Vdragon <mail@massdriver.space>

VynDragon commented:
@dpgeorge I added a couple new features that means it can replace flash area on rpi_pico, i've set it up to use this new thing.
For enabling the tests it would only be needed to add rpi_pico to the build list (I have zero familiarity with any of the other boards... Will give it an attempt if enabling rpi_pico in tests is not acceptable)

VynDragon committed:
zephyr/modules: Add Zephyr FileSystem support to _boot.py.

Signed-off-by: Vdragon <mail@massdriver.space>

VynDragon commented:
Added conversion from zephyr errno to mpy strings so mpy tools (mpremote, viper-ide, etc) works correctly

dpgeorge commented on ports/zephyr/modules/_boot.py:
```diff
@@ -54,7 +57,34 @@ def mount_all_disks():
 
 
-if FlashArea and create_flash_partition():
+def mount_FS_flash():
```
I suggest putting this function before `create_flash_partition()`, because that keeps the order they are called in.

dpgeorge commented on ports/zephyr/zephyr_fs.c:
```diff
@@ -0,0 +1,782 @@
+                break;
+
+            case 2:             // SEEK_END
+                err = fs_seek(&self->fp, s->offset, FS_SEEK_END);
```
Suggest simplifying this switch:
```c
MP_STATIC_ASSERT(FS_SEEK_SET == 0);
MP_STATIC_ASSERT(FS_SEEK_CUR == 1);
MP_STATIC_ASSERT(FS_SEEK_END == 2);
err = fs_seek(&self->fp, s->offset, s->whence);
```
Alternatively, use a lookup table:
```c
static const uint8_t WHENCE[] = {FS_SEEK_SET, FS_SEEK_CUR, FS_SEEK_END};
... WHENCE[s->whence]...
```

dpgeorge commented on ports/zephyr/zephyr_fs.c:
```diff
@@ -0,0 +1,782 @@
+    t->items[1] = MP_OBJ_NEW_SMALL_INT(stats.f_frsize);     // f_frsize
+    t->items[2] = MP_OBJ_NEW_SMALL_INT(stats.f_blocks);     // f_blocks
+    t->items[3] = MP_OBJ_NEW_SMALL_INT(stats.f_bfree);     // f_bfree
+    t->items[4] = t->items[3];     // f_bavail
```
I don't think these values are correct.  If I do `mpremote df` I get:
```
$ mpremote df
filesystem     size     used    avail use% mounted on
<FS>          90112      512    89600    0 /flash
```
But the same filesystem with `FlashArea` is:
```
$ mpremote df
filesystem     size     used    avail use% mounted on
<VfsLfs2>   1441792     8192  1433600    0 /flash
```
So the FS implementation here is a factor of 16 too small.

VynDragon replied:
What platform is that on? RP2040? It's very possible this is coming from different settings so would need to be compared to the settings of the DTS instead

VynDragon replied:
well added a if, but dont know what is happening on zephry side... done

dpgeorge replied:
> What platform is that on? RP2040?

Yes, sorry, should have mentioned it was a standard Pico with 2MiB flash, using the `rpi_pico` zephyr board definition from this repo.

josuah commented on ports/zephyr/zephyr_fs.c:
```diff
@@ -0,0 +1,782 @@
+    t->items[6] = MP_OBJ_NEW_SMALL_INT(0);     // f_ffree
+    t->items[7] = MP_OBJ_NEW_SMALL_INT(0);     // f_favail
+    t->items[8] = MP_OBJ_NEW_SMALL_INT(0);     // f_flags
+    t->items[9] = MP_OBJ_NEW_SMALL_INT(2048);     // f_namemax. Value i picked
```
https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/limits.h.html

- `POSIX` picked `PATH_MAX` to 256 bytes min
- `XOPEN` picked `PATH_MAX` to 1024 bytes min

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/uapi/linux/limits.h

- Linux picked `PATH_MAX` to 4096 bytes min

https://github.com/search?q=repo%3Azephyrproject-rtos%2Fzephyr%20%22%23define%20PATH_MAX%22&type=code

- Zephyr picks `PATH_MAX` to 256 bytes min, which can be used with `#include <limits.h>`

VynDragon replied:
is PATH_MAX equivalent to namemax? or better to use'_XOPEN_NAME_MAX' which is 255?

VynDragon commented on ports/zephyr/modzephyr.h:
```diff
@@ -37,4 +37,10 @@ extern const mp_obj_type_t zephyr_disk_access_type;
+#ifdef CONFIG_FILE_SYSTEM
+extern const mp_obj_type_t zephyr_fs_type;
+size_t mp_zephyr_fs_get_mounts_size(void);
+struct fs_mount_t *const *mp_zephyr_fs_get_mounts(void);
```
done

VynDragon commented on ports/zephyr/modzephyr.c:
```diff
@@ -86,6 +86,9 @@ static const mp_rom_map_elem_t mp_module_time_globals_table[] = {
     { MP_ROM_QSTR(MP_QSTR_FlashArea), MP_ROM_PTR(&zephyr_flash_area_type) },
     #endif
+    #ifdef CONFIG_FILE_SYSTEM
+    { MP_ROM_QSTR(MP_QSTR_FS), MP_ROM_PTR(&zephyr_fs_type) },
```
done

VynDragon commented:
Yea try adding CONFIG_FAT_FILESYSTEM_ELM=y and setting CONFIG_MICROPY_VFS_FAT=y in the .conf of rpi_pico, or removing the commit that enables disabling it.

dpgeorge commented on ports/zephyr/CMakeLists.txt:
```diff
@@ -55,6 +55,7 @@ set(MICROPY_SOURCE_PORT
     uart_core.c
     zephyr_device.c
     zephyr_storage.c
+    zephyr_fs.c
```
suggest calling the file `zephyr_filesystem.c` to match the type name

VynDragon commented on ports/zephyr/modules/_boot.py:
```diff
@@ -11,12 +11,39 @@
 _FLASH_EXISTS = False
 
 
+def mount_FS_flash():
```
done

dpgeorge reviewed (approved):
Thanks for updating, this is looking good now.
